azure/services/scalesetvms/client.go

```diff
@@ -74,79 +59,56 @@ func newVirtualMachineScaleSetVMsClient(subscriptionID string, baseURI string, a
 }
 
 // Get retrieves the Virtual Machine Scale Set Virtual Machine.
-func (ac *azureClient) Get(ctx context.Context, resourceGroupName, vmssName, instanceID string) (compute.VirtualMachineScaleSetVM, error) {
+func (ac *azureClient) Get(ctx context.Context, spec azure.ResourceSpecGetter) (interface{}, error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "scalesetvms.azureClient.Get")
 	defer done()
 
-	return ac.scalesetvms.Get(ctx, resourceGroupName, vmssName, instanceID, "")
+	return ac.scalesetvms.Get(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceName(), "")
 }
 
-// GetResultIfDone fetches the result of a long-running operation future if it is done.
-func (ac *azureClient) GetResultIfDone(ctx context.Context, future *infrav1.Future) (compute.VirtualMachineScaleSetVM, error) {
-	ctx, _, spanDone := tele.StartSpanWithLogger(ctx, "scalesetvms.azureClient.GetResultIfDone")
-	defer spanDone()
+// CreateOrUpdateAsync is a dummy implementation to fulfill the async.Reconciler interface.
```

Comment:
you shouldn't need to do this, you should be able to do ` async.New(scope, nil, client),` when creating the service in scalesetvms.go so this client doesn't need to implement the Deleter interface

check out the disks client for a similar example

Reply:
If I understand correctly, in the disks service we don't attempt to call `CreateOrUpdateResource()` so we can pass in a nil to the Creator interface. In this service, we want to get the resource if it exists and so I'm trying to leverage the `CreateOrUpdateResource()` in the async interface to fetch the resource and handle the not found/transient errors. Alternatively, I could try to construct the client as well, but I felt that it would be clunky to declare a Reconciler and a client for each type of VM.

Reply:
Now that I'm thinking about this more, I wonder if it might make more sense to add a `Getter` (https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/async/interfaces.go#L41)  in the Service and simply call `getter.Get()` in the `Reconcile` func instead of using `CreateOrUpdateResource` in a way it's not meant to be used.

Reply:
My thinking was to be able to take advantage of the error handling in [this block](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/c20099b2b691143f1034ef8b9309ed932c8e1b7e/azure/services/async/async.go#L116-L123) of `CreateOrUpdateResource()` so we don't have to duplicate that logic. Maybe we could add a `Get()` function to the `Reconciler` iface to give us the error handling?

Reply:
@mboersma what do you think? I'm concerned this won't work the same way with the async poller framework when we try to migrade scaleset vms to sdk v2

Reply:
Actually, I think this will act the same way in `asyncpoller`; the `CreateOrUpdateResource` code isn't really changed. I think we can merge this as-is and it won't present problems for refactoring SDKv2 on top of it.